File: go/test/ain2_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/sodeom-ai-proxy-sdk"
	"github.com/voxgig-sdk/sodeom-ai-proxy-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestAin2Entity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Ain2(nil)
		if ent == nil {
			t.Fatal("expected non-nil Ain2Entity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := ain2BasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "ain2." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set SODEOMAIPROXY_TEST_AIN__ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		ain2Ref01Ent := client.Ain2(nil)
		ain2Ref01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "ain2"}, setup.data), "ain2_ref01"))

		ain2Ref01DataResult, err := ain2Ref01Ent.Create(ain2Ref01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		ain2Ref01Data = core.ToMapAny(ain2Ref01DataResult)
		if ain2Ref01Data == nil {
			t.Fatal("expected create result to be a map")
		}

	})
}

func ain2BasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "ain2", "Ain2TestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read ain2 test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse ain2 test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"ain201", "ain202", "ain203"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("SODEOMAIPROXY_TEST_AIN__ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"SODEOMAIPROXY_TEST_AIN__ENTID": idmap,
		"SODEOMAIPROXY_TEST_LIVE":      "FALSE",
		"SODEOMAIPROXY_TEST_EXPLAIN":   "FALSE",
		"SODEOMAIPROXY_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["SODEOMAIPROXY_TEST_AIN__ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["SODEOMAIPROXY_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["SODEOMAIPROXY_APIKEY"],
			},
			extra,
		})
		client = sdk.NewSodeomAiProxySDK(core.ToMapAny(mergedOpts))
	}

	live := env["SODEOMAIPROXY_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["SODEOMAIPROXY_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/ain2_entity_test.rb
# Ain2 entity test

require "minitest/autorun"
require "json"
require_relative "../SodeomAiProxy_sdk"
require_relative "runner"

class Ain2EntityTest < Minitest::Test
  def test_create_instance
    testsdk = SodeomAiProxySDK.test(nil, nil)
    ent = testsdk.Ain2(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = ain2_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "ain2." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set SODEOMAIPROXY_TEST_AIN__ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    ain2_ref01_ent = client.Ain2(nil)
    ain2_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.ain2"), "ain2_ref01"))

    ain2_ref01_data_result, err = ain2_ref01_ent.create(ain2_ref01_data, nil)
    assert_nil err
    ain2_ref01_data = Helpers.to_map(ain2_ref01_data_result)
    assert !ain2_ref01_data.nil?

  end
end

def ain2_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "ain2", "Ain2TestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = SodeomAiProxySDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["ain201", "ain202", "ain203"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["SODEOMAIPROXY_TEST_AIN__ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "SODEOMAIPROXY_TEST_AIN__ENTID" => idmap,
    "SODEOMAIPROXY_TEST_LIVE" => "FALSE",
    "SODEOMAIPROXY_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["SODEOMAIPROXY_TEST_AIN__ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["SODEOMAIPROXY_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = SodeomAiProxySDK.new(Helpers.to_map(merged_opts))
  end

  live = env["SODEOMAIPROXY_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["SODEOMAIPROXY_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
